code/game/machinery/firealarm.dm

```diff
@@ -326,3 +326,109 @@
 	if (!party_overlay)
 		party_overlay = iconstate2appearance('icons/turf/areas.dmi', "party")
 	A.add_overlay(party_overlay)
+
+/obj/machinery/sprinkler
+	name = "foam dispenser"
+	desc = "An automated sprinkler cabable of detecting fire and spray coolant."
+	icon = 'icons/obj/device.dmi'
+	icon_state = "sprinkler0"
+	layer = PROJECTILE_HIT_THRESHHOLD_LAYER
+	plane = FLOOR_PLANE
+	max_integrity = 100
+	armor = list("melee" = 50, "bullet" = 20, "laser" = 20, "energy" = 20, "bomb" = 0, "bio" = 0, "rad" = 0, "fire" = 80, "acid" = 30)
+	var/obj/item/reagent_containers/glass/beaker/internal
+	var/refill_loaded = FALSE
+	req_access = list(ACCESS_ATMOSPHERICS)
+	var/cooldown = 0
+
+/obj/machinery/sprinkler/Initialize()
+	. = ..()
+	internal = new /obj/item/reagent_containers/glass/beaker(src)

```

Comment:
I suggest giving it its own `reagents` rather than using a completely separate container.